bundles/org.dataflowanalysis.analysis/src/org/dataflowanalysis/analysis/dsl/AnalysisConstraint.java

```diff
@@ -109,4 +135,79 @@ public void addConditionalSelector(ConditionalSelector selector) {
     public DSLContext getContext() {
         return context;
     }
+
+    @Override
+    public String toString() {
+        StringJoiner dslString = new StringJoiner(" ");
+        if (!this.dataSourceSelectors.getSelectors().isEmpty()) {
+            dslString.add(this.dataSourceSelectors.toString());
+        }
+        if (!this.vertexSourceSelectors.getSelectors().isEmpty()) {
+            dslString.add(this.vertexSourceSelectors.toString());
+        }
+        dslString.add(DSL_KEYWORD);
+        if (!this.vertexDestinationSelectors.getSelectors().isEmpty()) {
+            dslString.add(this.vertexDestinationSelectors.toString());
+        }
+        if (!this.conditionalSelectors.getSelectors().isEmpty()) {
+            dslString.add(this.conditionalSelectors.toString());
+        }
+        return dslString.toString();
+    }
+
+    public static ParseResult<AnalysisConstraint> fromString(StringView string) {
+    	return AnalysisConstraint.fromString(string, null);
+    }
+    
+    public static ParseResult<AnalysisConstraint> fromString(StringView string, DSLContextProvider contextProvider) {
+        DSLContext context = new DSLContext(contextProvider);
+        var sourceSelectors = parseSourceSelector(string, context);
+        if (sourceSelectors.failed()) {
+            return ParseResult.error(sourceSelectors.getError());
+        }
+        DataSourceSelectors dataSourceSelectors = sourceSelectors.getResult().getDataSourceSelectors().orElse(new DataSourceSelectors());
+        VertexSourceSelectors vertexSourceSelectors = sourceSelectors.getResult().getNodeSourceSelectors().orElse(new VertexSourceSelectors());
+
+        if (!string.startsWith(DSL_KEYWORD)) {
+            return string.expect(DSL_KEYWORD);
+        }
+        string.advance(DSL_KEYWORD.length() + 1);
+
+        ParseResult<VertexDestinationSelectors> nodeDestinationSelectorsParseResult = VertexDestinationSelectors.fromString(string, context);
+        if (nodeDestinationSelectorsParseResult.failed()) {
+            return ParseResult.error(nodeDestinationSelectorsParseResult.getError());
+        }
+        VertexDestinationSelectors vertexDestinationSelectors = nodeDestinationSelectorsParseResult.getResult();
+
+        ParseResult<ConditionalSelectors> conditionalSelectorsParseResult = ConditionalSelectors.fromString(string, context);
+        ConditionalSelectors conditionalSelectors = conditionalSelectorsParseResult.or(new ConditionalSelectors());
+
+        if (!string.empty()) {
+            return ParseResult.error("Unexpected symbols: " + string.getString());
+        }
+
+        return ParseResult.ok(new AnalysisConstraint(vertexSourceSelectors, dataSourceSelectors, vertexDestinationSelectors, conditionalSelectors, context));
+    }
+
+    public static ParseResult<SourceSelectors> parseSourceSelector(StringView string, DSLContext context) {
```

Comment:
Should this be public?

Reply:
Fixed in [baebd1e](https://github.com/DataFlowAnalysis/DataFlowAnalysis/pull/231/commits/baebd1ee610009d51427f3679bb32169ed644f34)